pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java

```diff
@@ -266,17 +271,35 @@ void messageAcked(CommandAck ack) {
         } else {
             subscription.acknowledgeMessage(position, ack.getAckType());
         }
+        
+        // unblock consumer-throttling when receives half of maxUnackedMessages => consumer can start again consuming
+        // messages
+        if (SubType.Shared.equals(subType)
+                && ((unackedMessages.decrementAndGet() == (maxUnackedMessages / 2)) && blockedConsumerOnUnackedMsgs)) {
+            blockedConsumerOnUnackedMsgs = false;
+            subscription.consumerFlow(this, messagePermits.get());
+            unackedMessages.addAndGet(messagePermits.get());
+        }
     }
 
     void flowPermits(int additionalNumberOfMessages) {
         checkArgument(additionalNumberOfMessages > 0);
 
+        // block shared consumer when unacked-messages reaches limit
+        if (SubType.Shared.equals(subType) && unackedMessages.get() >= maxUnackedMessages) {
+            blockedConsumerOnUnackedMsgs = true;
+        }
+
         int oldPermits = messagePermits.getAndAdd(additionalNumberOfMessages);
-        if (log.isDebugEnabled()) {
-            log.debug("[{}] Added more flow control message permits {} (old was: {})", this, additionalNumberOfMessages,
-                    oldPermits);
+        if (!blockedConsumerOnUnackedMsgs) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Added more flow control message permits {} (old was: {})", this,
+                        additionalNumberOfMessages, oldPermits);
+            }
+            subscription.consumerFlow(this, additionalNumberOfMessages);
+            unackedMessages.addAndGet(additionalNumberOfMessages);
```

Comment:
as we talked: we are not moving increment of unackedMessages at sendMessages(..) because this method is being called asynchronously and before it updates unackedMessages counter: if flowPermits(...) get triggered then it will add more permits and call dispatcher to dispatch more messages.